internal/cmd/gocdk/biome.go

```diff
@@ -246,17 +263,91 @@ func biomeApply(ctx context.Context, pctx *processContext, biome string, input b
 		return xerrors.Errorf("biome apply %s: %w", biome, err)
 	}
 
-	c := pctx.NewCommand(ctx, biomePath, "terraform", "init", "-input="+strconv.FormatBool(input))
-	if err := c.Run(); err != nil {
+	writeTerraformOutput := func(cmd *exec.Cmd, msg []byte) {
+		const separator = "#########################################\n"
+		pctx.stderr.Write([]byte("\n"))
+		pctx.stderr.Write([]byte(separator))
+		pctx.stderr.Write([]byte(fmt.Sprintf("Output from %q\n", strings.Join(cmd.Args, " "))))
+		pctx.stderr.Write([]byte(separator))
+		pctx.stderr.Write(msg)
+		pctx.stderr.Write([]byte(separator))
+		pctx.stderr.Write([]byte("\n"))
+	}
+
+	pctx.Logf("Checking to see if resource updates are needed...")
+
+	// First, run "terraform init". It's needed the first time, and anytime
+	// a new Terraform provider is added, so it's safest to just always run it.
+	inputArg := fmt.Sprintf("-input=%v", opts.Input)
+	c := pctx.NewCommand(ctx, biomePath, "terraform", "init", inputArg)
+	c.Env = overrideEnv(c.Env, "TF_IN_AUTOMATION=1")
+	c.Stdout, c.Stderr = nil, nil
+	if out, err := c.CombinedOutput(); err != nil {
+		writeTerraformOutput(c, out)
 		return xerrors.Errorf("biome apply %s: %w", biome, err)
+	} else if opts.Verbose {
+		writeTerraformOutput(c, out)
 	}
 
-	// TODO(#1821): take over steps (plan, confirm, apply) so we can
-	// dictate the messaging and errors. We should visually differentiate
-	// when we insert verbiage on top of terraform.
-	c = pctx.NewCommand(ctx, biomePath, "terraform", "apply", "-input="+strconv.FormatBool(input))
-	if err := c.Run(); err != nil {
+	// Next, run "terraform plan".
+	planFile := filepath.Join(biomePath, "tf.plan")
+	defer os.Remove(planFile)
+	c = pctx.NewCommand(ctx, biomePath, "terraform", "plan", "-detailed-exitcode", "-out="+planFile, inputArg)
+	c.Env = overrideEnv(c.Env, "TF_IN_AUTOMATION=1")
+	c.Stdout, c.Stderr = nil, nil
+
+	// No error means success and no diffs.
+	// Exit code 1 means the command failed.
+	// Exit code 2 means the command succeeded, but there's a diff.
+	if out, err := c.CombinedOutput(); err != nil && !strings.Contains(err.Error(), "exit status 2") {
+		// A real failure.
+		writeTerraformOutput(c, out)
 		return xerrors.Errorf("biome apply %s: %w", biome, err)
+	} else if err != nil {
+		// Exit code 2 --> there's a diff.
+		// Print them out and prompt for approval.
+		pctx.Logf("Resources updates are needed.")
+		if opts.Quiet {
+			pctx.Logf("Not showing update details due to --quiet.")
+		} else {
+			writeTerraformOutput(c, out)
```

Comment:
This prints all of the output from `terraform plan` -- as you can see from the example it is still quite chatty. We could try to parse it to reduce it.... Ick.

Reply:
As discussed offline, I think a better behavior would be to nix the `--quiet` flag, print out the summary of how many resources added, removed, changed by default, and then make it so that full Terraform output is shown when passing `--verbose`.

Reply:
Done.